File: files/2025-08-17/CSEC 17082025.py
import pandas as pd
import string

df1 = pd.read_excel("files/2025-08-17/Challenge 51.xlsx", usecols="B:D", skiprows=1, nrows=1)
df2 = pd.read_excel("files/2025-08-17/Challenge 51.xlsx", usecols="F:G", skiprows=1, nrows=6)

def col(i): return "" if i == 0 else col((i-1)//26) + string.ascii_uppercase[(i-1)%26]
def idx(x): return [col(i) for i in range(1,16385)].index(x)+1
def excel_cols(a,b): return [col(i) for i in range(idx(a),idx(b)+1)]

seqs = sum([excel_cols(r['From Column'], r['To Column']) for _,r in df1.iterrows()], [])
groups = [(i//3)+1 for i in range(len(seqs))]
res = (
    pd.DataFrame({'seq': seqs, 'GROUP': groups})
      .groupby('GROUP')['seq']
      .apply(lambda x: ','.join(x))
      .reset_index()
      .rename(columns={'seq': 'COLUMNS'})
)

print(res['COLUMNS'].equals(df2['COLUMNS'])) # True

File: files/2025-08-17/CSEC 17082025.R
library(tidyverse)
library(readxl)

path = "files/2025-08-17/Challenge 51.xlsx"
input = read_excel(path, range = "B2:D3")
test  = read_excel(path, range = "F2:G7")

excel_cols = function(start, end) {
  col = function(i) ifelse(i == 0, "", paste0(col((i - 1) %/% 26), LETTERS[(i - 1) %% 26 + 1]))
  idx = function(x) match(x, map_chr(1:16384, col))
  map_chr(seq(idx(start), idx(end)), col)
}

result = input %>%
  mutate(seq = map2(`From Column`, `To Column`, 
                   ~ excel_cols(.x, .y))) %>%
  unnest(seq) %>%
  mutate(GROUP = (row_number()-1) %/% 3 + 1) %>%
  summarise(COLUMNS = paste0(seq, collapse = ","),
            .by = GROUP) 

all.equal(result$COLUMNS, test$COLUMNS)
# > [1] TRUE